### PR TITLE
[Shopify] Log Skipped Records on export processes

### PR DESCRIPTION
**Pull request by tinestaric**
This pull request does not have a related issue as it's part of delivery for development agreed directly with @AndreiPanko 
Fixes #26819 

### Created new structure to log records skipped on export

Solution has new structure possible to open by typing "Shopify Skipped Records". On page user can see all the skipped records with reasons and date and time when the record was skipped. User can also open the related record using "Show record" action or clicking the "Description" field value.

### Log cases:

Customer:
- Customer has empty email
- Customer with same email or phone exists

Posted Sales Invoice:
- Customer does not exist in Shopify
- Payment terms do not exist in Shopify
- Customer No. is Default Customer No. for Shopify Shop
- Customer No. is used in Shopify Customer Template
- No lines existing in sales invoice
- Invalid Quantity
- Empty No. value

Product:
-  Item is blocked/sales blocked (Item Variant)
-  Item is blocked

Shipments:
 - Related Shopify Order does not exist
 - No lines in Posted Sales Shipment applicable for fulfillment 
 - No corresponding fulfillment found in Shopify.

If the condition met log record is being created.

### Retention policy
Retention policy is inserted on extension install. It can be turned on "Retention Policies" page.

Fixes [AB#473306](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/473306)

### CODE REVIEW

File: Apps/W1/Shopify/app/src/Customers/Codeunits/ShpfyCustomerExport.Codeunit.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Customers/Codeunits/ShpfyCustomerExport.Codeunit.al:93]: We have the exact logic here: https://github.com/microsoft/ALAppExtensions/blob/main/Apps/W1/Shopify/app/src/Companies/Codeunits/ShpfyCompanyExport.Codeunit.al

We should probably add it there too

File: Apps/W1/Shopify/app/src/Base/Codeunits/ShpfyInstaller.Codeunit.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Base/Codeunits/ShpfyInstaller.Codeunit.al:42]: This will not get executed unless a user invokes "Refresh Allowed Tables" because IsInitialSetup will always be false. GetShopifyLogEntryAddedToAllowedListUpgradeTag is already inserted into all upgraded databases.

We should add a new upgrade tag and move skipped records into a new procedure. Since you won't be able to (UpgradeTag is in Base Application), I will do this. Leaving the comment as a reminder to myself

File: Apps/W1/Shopify/app/src/Logs/Codeunits/ShpfySkipRecordMgt.Codeunit.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Logs/Codeunits/ShpfySkipRecordMgt.Codeunit.al:6]: Let's rename this to "Shpfy Skipped Record"

Mgt. does not seem very necessary

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductExport.Codeunit.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductExport.Codeunit.al:556]:
```suggestion
        ItemIsDraftLbl: Label 'Shopify product is in draft status.';
```

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductExport.Codeunit.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductExport.Codeunit.al:557]:
```suggestion
        ItemIsArchivedLbl: Label 'Shopify product is archived.';
```

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductExport.Codeunit.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductExport.Codeunit.al:68]:
```suggestion
        VariantPriceCalcSkippedLbl: Label 'Variant price is not synchronized because the item is blocked or sales blocked.';
```

File: Apps/W1/Shopify/app/src/Logs/Tables/ShpfySkippedRecord.Table.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Logs/Tables/ShpfySkippedRecord.Table.al:26]:
```suggestion
            Caption = 'Shopify Id';
```

File: Apps/W1/Shopify/app/src/Logs/Pages/ShpfySkippedRecords.Page.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Logs/Pages/ShpfySkippedRecords.Page.al:23]: I don't think this field adds any value

File: Apps/W1/Shopify/app/src/Logs/Pages/ShpfySkippedRecords.Page.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Logs/Pages/ShpfySkippedRecords.Page.al:6]: We should probably add these actions to this page:
image

@AndreiPanko please state if otherwise

File: Apps/W1/Shopify/app/src/Customers/Codeunits/ShpfyCustomerExport.Codeunit.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Customers/Codeunits/ShpfyCustomerExport.Codeunit.al:94]: This will log same record as skipped multiple times if user tries to sync multiple times. Should we do some duplicate handling? @AndreiPanko what do you think?

File: Apps/W1/Shopify/app/src/Invoicing/Codeunits/ShpfyPostedInvoiceExport.Codeunit.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Invoicing/Codeunits/ShpfyPostedInvoiceExport.Codeunit.al:99]:
```suggestion
        CustomerTemplateExistsLbl: Label 'Shopify customer template exists for customer no. %1 shop %2.', Comment = '%1 = Customer No., %2 = Shop Code';
```

File: Apps/W1/Shopify/app/src/Logs/Tables/ShpfySkippedRecord.Table.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Logs/Tables/ShpfySkippedRecord.Table.al:31]:
```suggestion
            Caption = 'Table Id';
```

File: Apps/W1/Shopify/app/src/Logs/Tables/ShpfySkippedRecord.Table.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Logs/Tables/ShpfySkippedRecord.Table.al:32]:
```suggestion
            ToolTip = 'Specifies the table Id of the skipped record.';
```

File: Apps/W1/Shopify/app/src/Logs/Tables/ShpfySkippedRecord.Table.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Logs/Tables/ShpfySkippedRecord.Table.al:43]:
```suggestion
            ToolTip = 'Specifies the table name of the skipped record.';
```

File: Apps/W1/Shopify/app/src/Logs/Tables/ShpfySkippedRecord.Table.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Logs/Tables/ShpfySkippedRecord.Table.al:48]:
```suggestion
            Caption = 'Record Id';
```

File: Apps/W1/Shopify/app/src/Logs/Tables/ShpfySkippedRecord.Table.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Logs/Tables/ShpfySkippedRecord.Table.al:49]:
```suggestion
            ToolTip = 'Specifies the record Id of the skipped record.';
```

File: Apps/W1/Shopify/app/src/Logs/Pages/ShpfySkippedRecords.Page.al
Comment (AndreiPanko) [reply Apps/W1/Shopify/app/src/Logs/Pages/ShpfySkippedRecords.Page.al:6]: Agree. Should be aligned

File: Apps/W1/Shopify/app/src/Companies/Codeunits/ShpfyCompanyExport.Codeunit.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Companies/Codeunits/ShpfyCompanyExport.Codeunit.al:52]:
```suggestion
            ShpfySkipRecord.LogSkippedRecord(Customer.RecordId, EmptyEmailAddressLbl, Shop);
```

File: Apps/W1/Shopify/app/src/Logs/Codeunits/ShpfySkipRecordMgt.Codeunit.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Logs/Codeunits/ShpfySkipRecordMgt.Codeunit.al:6]: Sorry to be a drag but isn't Shpfy Skipped Record a better name?

File: Apps/W1/Shopify/app/src/Companies/Codeunits/ShpfyCompanyExport.Codeunit.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Companies/Codeunits/ShpfyCompanyExport.Codeunit.al:40]: The naming of the variable is inconsistent. Let's use SkippedRecord (after renaming the codeunit).

Rule: Lose Shpfy or Shopify if same name does not exist in Base Application. Add Shopify if it exists.

Example:

"Shpfy Shop" -> Shop
"Shpfy Customer" -> ShopifyCustomer

File: Apps/W1/Shopify/app/src/Logs/Pages/ShpfySkippedRecords.Page.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Logs/Pages/ShpfySkippedRecords.Page.al:74]: I think this from copy paste. We should make the tooltip meaningful, not log entries but skipped record entries

File: Apps/W1/Shopify/app/src/Logs/Pages/ShpfySkippedRecords.Page.al
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Logs/Pages/ShpfySkippedRecords.Page.al:86]: I think this from copy paste. We should make the tooltip meaningful, not log entries but skipped record entries

File: Apps/W1/Shopify/app/src/Companies/Codeunits/ShpfyCompanyExport.Codeunit.al
Comment (petemchlk) [reply Apps/W1/Shopify/app/src/Companies/Codeunits/ShpfyCompanyExport.Codeunit.al:3]: strange, I guess it happened automatically somehow

Comment (tinestaric) [reply Apps/W1/Shopify/app/src/Companies/Codeunits/ShpfyCompanyExport.Codeunit.al:3]: Just a note for clarity, AZ AL Dev Tools adds the using when you define a variable. I opened a bug, that it shouldn't add the using if it matches the object namespace, but it's not yet fixed. 
So a note to @petemchlk, we have to be careful about that.

File: Apps/W1/Shopify/app/src/Base/Codeunits/ShpfyInstaller.Codeunit.al
Comment (onbuyuka) [reply Apps/W1/Shopify/app/src/Base/Codeunits/ShpfyInstaller.Codeunit.al:42]: added

File: Apps/W1/Shopify/app/src/Logs/Codeunits/ShpfySkipRecordMgt.Codeunit.al
Comment (onbuyuka) [reply Apps/W1/Shopify/app/src/Logs/Codeunits/ShpfySkipRecordMgt.Codeunit.al]: done

**Review by onbuyuka — APPROVED** [review]

File: Apps/W1/Shopify/app/src/Base/Codeunits/ShpfyInstaller.Codeunit.al
Comment (onbuyuka) [reply Apps/W1/Shopify/app/src/Base/Codeunits/ShpfyInstaller.Codeunit.al:42]: @JesperSchulz could you please add this comment to internal PR when you create it? You can resolve this comment when you create the PR

Comment (JesperSchulz) [reply Apps/W1/Shopify/app/src/Base/Codeunits/ShpfyInstaller.Codeunit.al:42]: @onbuyuka, I can see you created the comment yourself, hence resolving this one :-)

**Review by JesperSchulz — DISMISSED** [review]

**Comment by JesperSchulz** [comment]
The PR was successfully merged internally 🥳 This PR has a compilation issues which must get addressed before we can merge:
_AA0215 The file ShpfySkipRecordMgt.Codeunit.al has an incorrect name. The valid name is ShpfySkippedRecord.Codeunit.al._

**Comment by JesperSchulz** [comment]
Build still failing. Looks strange... The code is in, so it doesn't really matter. I would love to merge this PR though, rather than just close it. But it's also not worth spending time on. Your call :-)

**Pull request merged**
